march_rqt_gait_generator/src/march_rqt_gait_generator/gait_generator_controller.py

```diff
@@ -0,0 +1,310 @@
+            rospy.logdebug('Cannot start another time slider thread as one is already active')
+            return
+
+        current = self.view.time_slider.value()
+        playback_speed = self.playback_speed
+        max_time = self.view.time_slider.maximum()
+        self.time_slider_thread = self.view.create_time_slider_thread(current, playback_speed, max_time)
+        self.time_slider_thread.update_signal.connect(self.view.update_main_time_slider)
+        self.time_slider_thread.start()
+
+    def stop_time_slider_thread(self):
+        if self.time_slider_thread is not None:
+            self.time_slider_thread.stop()
+            self.time_slider_thread = None
+
+    def update_gait_duration(self, duration):
+        rescale_setpoints = self.view.scale_setpoints_check_box.isChecked()
+
+        if self.subgait.has_setpoints_after_duration(duration) and not rescale_setpoints:
+            if not self.subgait.has_multiple_setpoints_before_duration(duration):
+                self.view.message(title='Could not update gait duration',
+                                  msg='Not all joints have multiple setpoints before duration ' + str(duration))
+                self.view.duration_spin_box.setValue(self.subgait.duration)
+                return
+            discard_setpoints = self.view.yes_no_question(title='Gait duration lower than highest time setpoint',
+                                                          msg='Do you want to discard any setpoints higher than the '
+                                                          'given duration?')
+            if not discard_setpoints:
+                self.view.duration_spin_box.setValue(self.subgait.duration)
+                return
+        self.subgait.set_duration(duration, rescale_setpoints)
+        self.view.time_slider.setRange(0, 100 * self.subgait.duration)
+
+        was_playing = self.time_slider_thread is not None
+        self.stop_time_slider_thread()
+
+        self.view.update_joint_widgets(self.subgait.joints)
+
+        if was_playing:
+            self.start_time_slider_thread()
+
+    def import_gait(self):
+        file_name, f = self.view.open_file_dialogue()
+
+        gait = ModifiableSubgait.from_file(self.robot, file_name, self)
+        if gait is None:
+            rospy.logwarn('Could not load gait %s', file_name)
+            return
+        self.subgait = gait
+        self.view.load_gait_into_ui(self.subgait)
+        for joint in self.subgait.joints:
+            self.connect_plot(joint)
+        self.current_time = 0
+
+        self.gait_directory = '/'.join(file_name.split('/')[:-3])
+        rospy.loginfo('Setting gait directory to %s', str(self.gait_directory))
+        self.view.change_gait_directory_button.setText(self.gait_directory)
+
+        # Clear undo and redo history
+        self.joint_changed_history = RingBuffer(capacity=100, dtype=list)
+        self.joint_changed_redo_list = RingBuffer(capacity=100, dtype=list)
```

Comment:
I would make this method accept a file name and return the subgait and do all the file dialogs and changing stuff in the view in the `View` class.